Add warning when keys are mismatched

mvnnn opened the pull request:
fixes #4508

embray commented:
Apparently this is attempting to address #4508 (please mention this in your PR description).

embray commented on astropy/io/ascii/ui.py:
```diff
@@ -661,11 +664,25 @@ def write(table, output=None,  format=None, Writer=None, fast_writer=True, **kwa
+    if kwargs.get('formats') and kwargs.get('names'):
+        for key in kwargs['formats']:
+            if not key in kwargs['names']:
+                _diffkeyArray.append(key)
```
You an do this with sets:
```
names = set(kwargs.get('names', []))
formats = set(kwargs.get('formats', []))
diff = formats - names
if diff:
     # Blah blah blah, the following format keys do not match any of the given column names
```

embray commented on astropy/io/ascii/ui.py:
```diff
@@ -32,7 +33,9 @@
 from ...extern import six
+from ...utils.exceptions import AstropyDeprecationWarning
 
+_diffkeyArray=[]
```
There's no reason for this to be a global variable. That said, see https://github.com/astropy/astropy/pull/4511/files#r50255334

embray commented on astropy/io/ascii/ui.py:
```diff
@@ -661,11 +664,25 @@ def write(table, output=None,  format=None, Writer=None, fast_writer=True, **kwa
+    if len(_diffkeyArray) != 0:
+        warnings.warn(
+            '%r keys in the formats dictionary not to match the keys in names dictionary'
+            %_diffkeyArray, AstropyDeprecationWarning)
```
Here you would use just `AstropyWarning`--this is not a deprecation.  Also use `.format()` instead of old-style `%` formatting (this is still inconsistent in some areas but new style formatting should be used for new code)

embray commented on CHANGES.rst:
```diff
@@ -462,6 +462,10 @@ New Features
 
+  - if the keys in the formats dictionary not to match the names in the names
+    dictionary then it's raises warning to show which keys are not to match in
+    keys in names Dictionary. [#4508]
```
Thanks for adding a changelog entry.  However, it's in the wrong section--you've placed it in an existing release instead of one of the upcoming ones.  In any case I'm not sure yet what release this should be fixed in (I suppose a patch release would be fine probably though).

embray replied:
This will also have to be rewritten, I'm afraid.

taldcroft commented:
Unfortunately it is a bit trickier than this to actually get the final list of names that will be written out.  That is because you need to account for the input data column names and `names` and `include_names` and `exclude_names`.  So to do this here one needs to replicate [this logic](https://github.com/astropy/astropy/blob/4576274ca0bd2d74f6ec2775e17d8fb4e54733fa/astropy/io/ascii/core.py#L922).

mvnnn commented:
Oh, I will fix that.

mvnnn commented:
@taldcroft  i fixed that.

mvnnn commented:
@taldcroft  i fixed that and it's passed all checks but now my previous method as @embray suggested to me it's passed all checks (previously it's fail in one check).now i make which method's PR ?

taldcroft commented:
@mvnnn - the implementation here is still mostly incorrect and untested.  

I would suggest that you start by writing a substantial set of tests that cover many different options for `ascii.write(data, names=.., include_names=.., exclude_names=.., formats=...).  Once you have all those tests written, then try to write the code that makes them all pass locally on your machine (`python setup.py test --package=io.ascii`).  If you get stuck you can push up some code, but be sure to include `[skip ci]` in the commit message to avoid wasting travis CI time if you know it will fail.

mvnnn commented:
thanks @taldcroft . i will make new file in ("astropy/io/ascii/tests/test_arg_write.py") that can test different options for ascii.write( ).

mvnnn commented:
@taldcroft i added set of tests if any other test case is possible then tell me i added those. now i will make one function in ```io/ascii/ui.py```. that function can check all possible set of tests which i added.

taldcroft commented on astropy/io/ascii/ui.py:
```diff
@@ -661,6 +695,8 @@ def write(table, output=None,  format=None, Writer=None, fast_writer=True, **kwa
     Writer : ``Writer``
         Writer class (DEPRECATED) (default= :class:`Basic`)
     """
+    _check_any_invalide_arg(table, kwargs)
```
This needs to go after the line setting `table = Table(...)`.

taldcroft commented:
You need to replicate the logic https://github.com/astropy/astropy/blob/4576274ca0bd2d74f6ec2775e17d8fb4e54733fa/astropy/io/ascii/core.py#L922 almost exactly with the exception that you don't actually change the contents of `table`.  In particular the most common situation is that `names is None` and the names come from the table columns.

mvnnn commented:
@taldcroft if i rename table column names as you suggests above then it's makes ```AssertionError``` in ```test_write.py```

mvnnn commented:
@taldcroft i added condition for ```names is None``` but still something wrong.

mvnnn commented:
Does this look OK .. @taldcroft @embray

mvnnn commented:
Does this look good.@taldcroft @embray

mvnnn commented:
Does this look good.@taldcroft

mvnnn commented:
@taldcroft please check the changes.

mvnnn commented:
please check the changes.

mvnnn commented:
#4567

taldcroft commented:
@mvnnn - sorry for the delay, things are busy...

I would prefer that this PR stay within the confines of the original issue and *only* check that the `formats` keys are contained in the final set of names.

Can you explore a simpler implementation like this (within `ui.write(table, ..)`) and see if it can work?  I have not tried this out.
```
    if output is None:
        output = sys.stdout

    table = Table(table, names=kwargs.get('names'))

    # Check that all formats keywords are valid output column names
    table0 = table[:0].copy()  # Make a zero-length copy
    core._apply_include_exclude_names(table0, kwargs.get('names'), 
                    kwargs.get('include_names'), kwargs.get('exclude_names'))
    if any(fmt not in table0.colnames for fmt in kwargs.get('formats', [])):
        # warning
```

mvnnn commented:
@taldcroft please check the changes.

embray commented on CHANGES.rst:
```diff
@@ -776,6 +776,14 @@ API changes
+    to match in names Dictionary. if any keys in the exclude_names dictionary
+    match to any keys in include_names dictionary then it's make warning to show
+    which keys in the exclude_names dictionary are match in include_names
+    dictionary. [#4508]
```
This changelog entry will still need to be rewritten.  It really doesn't need to be so verbose, and it's also lacking context.    Something like 
```
Better validation of ``formats``, ``include_names``, and ``exclude_names`` arguments to ``ascii.write()``.
```
The details of this change really aren't deeply important from a user-perspective, and the issue reference allows someone to read up more on what is meant by this if they really care.

taldcroft commented on astropy/io/ascii/ui.py:
```diff
@@ -661,11 +663,24 @@ def write(table, output=None,  format=None, Writer=None, fast_writer=True, **kwa
+    table0 = table[:0].copy()
+    core._apply_include_exclude_names(table0, kwargs.get('names'),
+                    kwargs.get('include_names'), kwargs.get('exclude_names'))
+    if any(fmt not in table0.colnames for fmt in kwargs.get('formats', [])):
```
This `if` statement is now redundant, so just write the whole thing as:
```
    diff_format_with_names = set(kwargs.get('formats', [])) - set(table0.colnames)
    if diff_format_with_names:
        warnings.warn(...)
```

mvnnn replied:
that's change fail to pass ```Travis CI```

taldcroft commented on astropy/io/ascii/tests/test_arg_write.py:
```diff
@@ -0,0 +1,43 @@
+def test_name_with_formats(names, formats, output):
+    with catch_warnings(AstropyWarning) as ASwarn:
+        out = StringIO()
+        ascii.write(_data, out, names=names, formats=formats)
```
This doesn't test the most common case that the table (`_data` here) is defining the column names.

taldcroft commented on CHANGES.rst:
```diff
@@ -776,6 +776,10 @@ API changes
     for parsing a file with blank lines in the header containing whitespace will
     need to be updated. [#2654]
 
+  - if any keys in the formats dictionary are not to match any keys in the names
```
Sentences must always be capitalized (here and in code comments).  Along the lines that @embray already suggested, try:
```
 - Check that columns in `formats` specifier exist in the output table when writing. [#4508]
```

taldcroft commented on astropy/io/ascii/tests/test_arg_write.py:
```diff
@@ -0,0 +1,43 @@
+import warnings
```
There is no real need to create an entirely new test file for this one small case.  This can be put into `test_write.py`.

taldcroft commented on astropy/io/ascii/tests/test_arg_write.py:
```diff
@@ -0,0 +1,43 @@
+    (['x', 'y'], {'y':'%f'}, False),
+    (['x', 'y'], {}, False)
+])
+def test_name_with_formats(names, formats, output):
```
Put the comments above here as the test function docstring.

mvnnn commented:
this code throw same error as my first commit get  (https://github.com/mvnnn/astropy/commit/24ab57faaeb1e97d748fd0b3ef95d9b31d7c814d) .
but my second commit didn't get this error.
(https://github.com/mvnnn/astropy/commit/14d752910d71b3727a9e7c57bea6c61a6ad64877)

mvnnn commented:
@taldcroft please check the changes.

mvnnn commented:
@taldcroft please check the changes.

taldcroft commented:
Your tests still don't test the most common case of passing a `Table` object to `ascii.write`.

mvnnn commented:
@taldcroft please check the changes.

mvnnn commented:
@taldcroft please check the changes.

embray commented on CHANGES.rst:
```diff
@@ -811,6 +811,9 @@ API changes
     for parsing a file with blank lines in the header containing whitespace will
     need to be updated. [#2654]
 
+  - Check that columns in `formats` specifier exist in the output table
```
This should be
```
``formats``
```
with double-backquotes.

mvnnn commented:
@taldcroft please check the changes.

taldcroft commented on astropy/io/ascii/ui.py:
```diff
@@ -666,6 +668,26 @@ def write(table, output=None,  format=None, Writer=None, fast_writer=True, **kwa
+            '{} keys in the formats not to match the keys in names'
+            .format(diff_format_with_names), AstropyWarning)
+
+    if same_include_names_with_exclude_names:
```
I previously asked you to limit the scope to purely checking the format keys against the final output keys.  Please remove this.

mvnnn replied:
@taldcroft but previously you told me that to include```include_names```and```exclude_names``` test cases and check  if include_names = ['x','y'] and exclude_names=['y'] then show output 'x' is OK buy 'y' is not OK. so i added this.

mvnnn replied:
@taldcroft but previously you told me that to include ```include_names``` and ```exclude_names``` test cases and check if include_names = ['x','y'] and exclude_names=['y'] then show output 'x' is OK buy 'y' is not OK. so i added this. should i remove this ?

mvnnn replied:
@taldcroft but previously you told me that to include include_names and exclude_names test cases and check if include_names = ['x','y'] and exclude_names=['y'] then show output 'x' is OK buy 'y' is not OK. so i added this.

mvnnn replied:
@taldcroft  please review.

mvnnn commented:
@taldcroft but previously you told me that to include include_names and exclude_names test cases and check if include_names = ['x','y'] and exclude_names=['y'] then show output 'x' is OK buy 'y' is not OK. so i added this.

taldcroft commented:
@mvnnn - patience please.

saimn commented:
@mvnnn: posting 4 times the same comment will not make things go faster, and can be irritating for all people receiving notifications.

saimn commented on astropy/io/ascii/tests/test_write.py:
```diff
@@ -534,3 +535,46 @@ def test_byte_string_output(fast_writer):
+    t = table.Table([[1,2,3],[4.1,5.2,6.3]])
+    with catch_warnings(AstropyWarning) as ASwarn:
+        out = StringIO()
+        ascii.write(t, out, names=names, include_names=include_names, \
```
No need for backslash inside parenthesis.

saimn commented on astropy/io/ascii/tests/test_write.py:
```diff
@@ -534,3 +535,46 @@ def test_byte_string_output(fast_writer):
+        out = StringIO()
+        ascii.write(t, out, names=names, include_names=include_names, \
+        exclude_names=exclude_names, formats=formats)
+    assert (issues_warning == (len(ASwarn)>0))
```
This supposes that multiple warnings can be raised, is it the case ? If yes, the test could pass with a warning which is not the one you expect, so it would require a more specific test. But here I guess you can test that  `len(ASwarn) == 1`, or put the number of expected warnings in the fixture.

saimn commented on astropy/io/ascii/tests/test_write.py:
```diff
@@ -534,3 +535,46 @@ def test_byte_string_output(fast_writer):
+    value false and Check that columns in formats specifier exist in the
+    issues_warning table when writing.
+    """
+    t = table.Table([[1,2,3],[4.1,5.2,6.3]])
```
You could add `names=...` here to define some column names and have a more concrete test.

mvnnn replied:
Above test cases with name and below test case without name because in above test cases i focus to check ```formats``` with ```names``` but in below test cases i focus to check ```formats``` with ```table.colnames``` so we don't need to add ```names``` in below test cases.

saimn commented on astropy/io/ascii/ui.py:
```diff
@@ -666,6 +668,26 @@ def write(table, output=None,  format=None, Writer=None, fast_writer=True, **kwa
+
+    if diff_format_with_names:
+        warnings.warn(
+            '{} keys in the formats not to match the keys in names'
```
Maybe something like: `The keys "{}" specified in the 'formats' argument does not match a column name.` ?

mvnnn commented:
I apologies for the inconvenience. i don't know that all people receiving notifications.i will take care of that thing.

taldcroft commented on astropy/io/ascii/ui.py:
```diff
@@ -666,6 +668,26 @@ def write(table, output=None,  format=None, Writer=None, fast_writer=True, **kwa
+    core._apply_include_exclude_names(table0, kwargs.get('names'),
+                    kwargs.get('include_names'), kwargs.get('exclude_names'))
+    diff_format_with_names = set(kwargs.get('formats', [])) - set(table0.colnames)
+    same_include_names_with_exclude_names = set(kwargs.get('include_names', []))
```
Please remove these two (`same_include_...` and `diff_include_...`) and the `if same_include_..` below.  The original issue was to validate `formats` given any set of `names`, `include_names`, `exclude_names` inputs.  We do not require separate consistency of `include_names` and `exclude_names` since they do have a well-defined precedence order so overlapping is OK.

mvnnn commented:
Is it alright now?

taldcroft commented:
@mvnnn - merged.  Thanks!
